rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingPropertyTest.scala

```diff
@@ -1,27 +1,115 @@
+import coop.rchain.rspace.history.Branch
+import coop.rchain.rspace.{Context, RSpace}
+import monix.eval.Task
 import org.scalacheck.Arbitrary
 import org.scalacheck.Test.Parameters
 import org.scalatest.{FlatSpec, Matchers}
 import org.scalatest.prop.PropertyChecks
 
+import scala.concurrent.duration._
+import monix.execution.Scheduler.Implicits.global
+
 class CostAccountingPropertyTest extends FlatSpec with PropertyChecks with Matchers {
+  import CostAccountingPropertyTest._
+
   implicit val params: Parameters = Parameters.defaultVerbose.withMinSuccessfulTests(1000)
   implicit val procArbitrary = Arbitrary(
     ProcGen.topLevelGen(5).map(PrettyPrinted[Proc](_, PrettyPrinter.print))
   )
 
+  implicit val taskExecutionDuration = 5.seconds
+
   def cost(proc: Proc): Cost = Cost(Interpreter.buildPar(proc).apply)
 
   behavior of "Cost accounting in Reducer"
 
-  it should "cost of substitution is never negative" in {
+  it should "have positive cost of substitution" in {
     forAll { p: PrettyPrinted[Proc] =>
       cost(p.value).value should be >= 0L
     }
+  }
+
+  behavior of "Charging phlos in Reducer"
+
+  it should "charge the same if the evaluation order is reversed" in {
+    forAll { (p: PrettyPrinted[Proc], q: PrettyPrinted[Proc]) =>
+      haveEqualResults(costOfExecution(p.value, q.value), costOfExecution(q.value, p.value))
+    }
+  }
 
+  it should "sequential execution and parallel execution have the same cost" in {
+    forAll { (p: PrettyPrinted[Proc], q: PrettyPrinted[Proc]) =>
+      haveEqualResults(
+        costOfExecution(new PPar(p.value, q.value)),
+        costOfExecution(q.value, p.value)
+      )
+    }
   }
 }
+
+object CostAccountingPropertyTest {
+  def haveEqualResults[A](tasks: Task[A]*)(implicit duration: Duration): Boolean =
+    tasks.toList
+      .sequence[Task, A]
+      .map { _.sliding(2).forall { case List(r1, r2) => r1 == r2 } }
+      .runSyncUnsafe(duration)
+
+  def createRhoISpace(): RhoISpace = {
+    implicit val syncF: Sync[Id] = coop.rchain.catscontrib.effect.implicits.syncId
+    import coop.rchain.rholang.interpreter.storage.implicits._
+    val dbDir               = Files.createTempDirectory("cost-accounting-property-test-")
+    val context: RhoContext = Context.create(dbDir, 1024L * 1024L * 4)
+    val space: RhoISpace = RSpace.create[
+      Id,
+      Par,
+      BindPattern,
+      OutOfPhlogistonsError.type,
+      ListParWithRandom,
+      ListParWithRandomAndPhlos,
+      TaggedContinuation
+    ](context, Branch("test"))
+    space
+  }
+
+  def execute(reducer: ChargingReducer[Task], p: Proc)(
+      implicit rand: Blake2b512Random
+  ): Task[Long] = {
+    val program = Interpreter.buildPar(p).apply
+
+    val initPhlos = Cost(accounting.MAX_VALUE)
+
+    for {
+      _         <- reducer.setAvailablePhlos(initPhlos)
+      _         <- reducer.inj(program)
+      phlosLeft <- reducer.getAvailablePhlos()
+    } yield (initPhlos - phlosLeft.cost).value
```

Comment:
Why can't we reuse Interpreter.evaluate in here? I see two possible reasons, both questionable:
1. it does checkpoints - just extract the checkpointing away
2. it checks for evaluation error in the `errorVector` - which we should probably do too in this test? I mean: we want the tested scenarios to error at the same times, and there should probably be no errors in the first place.

Reply:
I used the example from CostAccountingReducerTest but, indeed, it would be better to use Interpreter.evaluate directly

Reply:
`CostAccountingReducerTest` needs to control the init phlos limit so it doesn't reuse the `interpreter.evaluate`.

Reply:
Then evaluate should have an overload taking the limit as parameter :)

Reply:
let's do this as a future refactoring step https://rchain.atlassian.net/browse/RHOL-1018